apps/annotation/fields.py

```diff
@@ -0,0 +1,184 @@
+    id_class = IDField
+
+    def __init__(self, resource_name, **kwargs):
+        self.type_subfield = self.type_class(const_value=resource_name)
+        self.type_subfield.bind(field_name='', parent=self)
+        self.id_subfield = self.id_class()
+        self.id_subfield.bind(field_name='', parent=self)
+        super().__init__(**kwargs)
+
+    def to_internal_value(self, data):
+        self.type_subfield.run_validation(data.get('type', None))
+        value = self.id_subfield.run_validation(data.get('id', None))
+        return value
+
+    def to_representation(self, value):
+        return {
+            'type': self.type_subfield.to_representation(None),
+            'id': self.id_subfield.to_representation(value)
+        }
+
+
+# Universal constant for marking the value of none relation (which does not mean whole output of field is =None).
+# It does not change reference even if called (as class-based const does).
+relation_none = lambda: relation_none
+
+
+class RelationField(serializers.Field):
+    child = _UnvalidatedField()
+    link_child_class = URLNameField
+
+    default_error_messages = {
+        'data_missing': "'data' key is missing",
+        'not_a_dict': _('Expected a dictionary of items but got type "{input_type}".'),
+    }
+
+    def __init__(self, related_link_url_name, many=False, **kwargs):
+        self.related_link_url_name = related_link_url_name
+        self.child = kwargs.pop('child', copy.deepcopy(self.child))
+        self.link_child = self.link_child_class(url_name=related_link_url_name)
+        self.many = many
+
+        if self.many:
+            self.child = serializers.ListField(child=self.child)
+        self.child.bind(field_name='', parent=self)
+        self.link_child.bind(field_name='', parent=self)
+        super().__init__(**kwargs)
+
+    def get_default(self):
+        default = super().get_default()
+        # relation_none -> None
+        # The internal value for relation_none is None.
+        if default is relation_none:
+            return [] if self.many else None
+        return default
+
+    def get_attribute(self, instance):
+        attribute = super().get_attribute(instance)
+        # None -> relation_none
+        # Attribute is not missing, but it's None, so we want to set none-relation instead, which outputs more than None
+        # value. To do that we override attribute with `relation_none`, so that we can add proper representation later.
+        if attribute is None:
+            return relation_none
+        return attribute
+
+    def to_internal_value(self, data):
+        if not isinstance(data, dict):
+            self.fail('not_a_dict', input_type=type(data).__name__)
+        if 'data' not in data:
+            self.fail('data_missing')
+        return self.child.run_validation(data['data'])
+
+    def to_representation(self, instance):
+        root_obj_id = getattr(self.context['root_resource_obj'], 'id', None) or self.context['root_resource_obj']
```

Comment:
I guess it's easy to forget `root_resource_obj` since, as far as I understand, it's our custom value we pass to the serializer via `context`. An assert/exception could be welcome ;)

Reply:
Sure :)